core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs

```diff
@@ -0,0 +1,249 @@
+        Self::Postgres(value)
+    }
+}
+
+impl<'a> From<RocksdbStorage> for PgOrRocksdbStorage<'a> {
+    fn from(value: RocksdbStorage) -> Self {
+        Self::Rocksdb(value)
+    }
+}
+
+/// A [`ReadStorageFactory`] implementation that can produce short-lived [`ReadStorage`] handles
+/// backed by either Postgres or RocksDB (if it's caught up). Always initialized as a `Postgres`
+/// variant and is then mutated into `Rocksdb` once RocksDB cache is caught up. After which it
+/// can never revert back to `Postgres` as we assume RocksDB cannot fall behind under normal state
+/// keeper operation.
+#[derive(Debug, Clone)]
+pub struct AsyncRocksdbCache {
+    pool: ConnectionPool,
+    rocksdb_cell: Arc<OnceCell<RocksDB<StateKeeperColumnFamily>>>,
+}
+
+impl AsyncRocksdbCache {
+    /// Load latest sealed miniblock from the latest sealed L1 batch (ignores sealed miniblocks
+    /// from unsealed batches).
+    async fn load_latest_sealed_miniblock(
+        connection: &mut StorageProcessor<'_>,
+    ) -> anyhow::Result<Option<(MiniblockNumber, L1BatchNumber)>> {
+        let mut dal = connection.blocks_dal();
+        let Some(l1_batch_number) = dal.get_sealed_l1_batch_number().await? else {
+            return Ok(None);
+        };
+        let Some((_, miniblock_number)) =
+            dal.get_miniblock_range_of_l1_batch(l1_batch_number).await?
+        else {
+            return Ok(None);
+        };
+        Ok(Some((miniblock_number, l1_batch_number)))
+    }
+
+    /// Returns a [`ReadStorage`] implementation backed by Postgres
+    async fn access_storage_pg(
+        rt_handle: Handle,
+        pool: &ConnectionPool,
+    ) -> anyhow::Result<PgOrRocksdbStorage<'_>> {
+        let mut connection = pool.access_storage().await?;
+
+        let (miniblock_number, l1_batch_number) =
+            match Self::load_latest_sealed_miniblock(&mut connection).await? {
+                Some((miniblock_number, l1_batch_number)) => (miniblock_number, l1_batch_number),
+                None => {
+                    tracing::info!("Could not find latest sealed miniblock, loading from snapshot");
+                    let snapshot_recovery = connection
+                        .snapshot_recovery_dal()
+                        .get_applied_snapshot_status()
+                        .await
+                        .context("failed getting snapshot recovery info")?
+                        .ok_or_else(|| {
+                            anyhow::anyhow!("Could not find snapshot, no state available")
+                        })?;
+                    (
+                        snapshot_recovery.miniblock_number,
+                        snapshot_recovery.l1_batch_number,
+                    )
+                }
+            };
+
+        tracing::debug!(%l1_batch_number, %miniblock_number, "Using Postgres-based storage");
+        Ok(
+            PostgresStorage::new_async(rt_handle, connection, miniblock_number, true)
```

Comment:
Nit: IMO, it wouldn't be controversial to retrieve a `Handle` here via [`Handle::current()`](https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.current), or even in `PostgresStorage::new_async()` so that it doesn't need to be passed around.